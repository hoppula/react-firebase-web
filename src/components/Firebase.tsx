import * as firebase from "firebase"
import * as PropTypes from "prop-types"
import * as React from "react"

export interface FirebaseProps {
  readonly apiKey: string
  readonly projectId: string
  readonly databaseURL?: string
}

export class Firebase extends React.Component<FirebaseProps, {}> {
  static propTypes = {
    apiKey: PropTypes.string.isRequired,
    databaseURL: PropTypes.string,
    projectId: PropTypes.string.isRequired
  }

  static childContextTypes = {
    firebase: PropTypes.object
  }

  firebase: firebase.app.App

  constructor(props: FirebaseProps) {
    super(props)
    const { apiKey, projectId, databaseURL } = props
    this.firebase = firebase.initializeApp({
      apiKey,
      authDomain: `${projectId}.firebaseapp.com`,
      databaseURL: databaseURL
        ? databaseURL
        : `https://${projectId}.firebaseio.com`,
      storageBucket: `${projectId}.appspot.com`
    })
  }

  getChildContext() {
    return {
      firebase: this.firebase
    }
  }

  render() {
    return React.Children.only(this.props.children)
  }
}

export default Firebase
