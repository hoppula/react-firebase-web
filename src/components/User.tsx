import * as firebase from "firebase"
import * as PropTypes from "prop-types"
import * as React from "react"
import { FirebaseContext } from "../types"

export interface UserProps {
  readonly mapUser?: (user: any) => any
  readonly children: (user: any) => React.ReactElement<any>
}

export interface UserState {
  user: any
}

export class User extends React.Component<UserProps, UserState> {
  static propTypes = {
    mapAuth: PropTypes.func,
    children: PropTypes.func.isRequired
  }

  static contextTypes = {
    firebase: PropTypes.object.isRequired
  }

  unsubscribe: firebase.Unsubscribe

  context: FirebaseContext
  state = {
    user: {}
  }

  componentDidMount() {
    const { mapUser } = this.props
    this.unsubscribe = this.context.firebase
      .auth()
      .onAuthStateChanged((authData: firebase.User) => {
        this.setState({
          user:
            mapUser && authData
              ? mapUser(authData.toJSON())
              : authData ? authData.toJSON() : {}
        })
      })
  }

  componentWillUnmount() {
    this.unsubscribe()
  }

  render() {
    return React.Children.only(this.props.children(this.state.user))
  }
}

export default User
