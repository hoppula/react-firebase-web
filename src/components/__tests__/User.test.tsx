import React from "react"
import renderer from "react-test-renderer"
import { Firebase } from "../Firebase"
import User from "../User"

import { initializeMockSDK } from "./setup"
const mocksdk = initializeMockSDK()

jest.mock("firebase/app", () => {
  return {
    initializeApp: (...args: any[]) => mocksdk.initializeApp(...args)
  }
})

describe("User", () => {
  test("renders different content when logged in", () => {
    let component = renderer.create(
      <Firebase apiKey="firebase-test" projectId="firebase-test">
        <User
          children={user => {
            if (user.uid) {
              return <div>{user.displayName} Logout</div>
            } else {
              return <div>Login</div>
            }
          }}
        />
      </Firebase>
    )
    const auth = component.getInstance().firebase.auth()

    // no user logged in
    let tree = component.toJSON()
    expect(tree).toMatchSnapshot()

    auth.changeAuthState({
      toJSON: () => ({ uid: "test", displayName: "Test User" })
    })
    auth.flush()

    // test user logged in
    tree = component.toJSON()
    expect(tree).toMatchSnapshot()
  })
})
