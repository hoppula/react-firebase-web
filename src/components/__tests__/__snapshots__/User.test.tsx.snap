// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`User renders different content when logged in 1`] = `
<div>
  Login
</div>
`;

exports[`User renders different content when logged in 2`] = `
<div>
  Test User
   Logout
</div>
`;
